### PR TITLE
Add roles support

### PR DESCRIPTION
**Pull request by alexhung**
For #1.

Add support for managing project roles. Only support custom role (i.e. `type="CUSTOM"`). Predefined roles can't be altered.

### CODE REVIEW

File: pkg/projects/resource_project.go [hunk 1]
```diff
@@ -177,18 +177,58 @@ func projectResource() *schema.Resource {
...
+						)),
+					},
+					"description": {
+						Type:     schema.TypeString,
+						Optional: true,
+					},
+					"type": {
+						Type:             schema.TypeString,
+						Required:         true,
+						ValidateDiagFunc: validation.ToDiagFunc(validation.StringMatch(customRoleTypeRegex, fmt.Sprintf("Only \"%s\" is supported", customRoleType))),
```
Comment (chb0github) [inline pkg/projects/resource_project.go:201]: in the future, if you have `""` in the string you can use the  gravemark string and then you don't need to escape

File: pkg/projects/resource_project_role.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+	Type         string   `json:"type"`
+	Environments []string `json:"environments"`
+	Actions      []string `json:"actions"`
+}
+
+func (r Role) Id() string {
+	return r.Name
+}
+
+func (a Role) Equals(b Identifiable) bool {
```
Comment (chb0github) [inline pkg/projects/resource_project_role.go:65]: I would say that you can apply this one any identifiable

File: pkg/projects/resource_project_role.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	var equatables []Equatable
+
+	for _, role := range roles {
+		equatables = append(equatables, role)
+	}
+
+	return equatables
+}
+
+func equatablesToRoles(equatables []Equatable) []Role {
```
Comment (chb0github) [inline pkg/projects/resource_project_role.go:79]: God, I thought we were past this shit when Java 8 came out 10 years ago

File: pkg/projects/resource_project_role.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		return nil, err
+	}
+
+	log.Printf("[TRACE] roles: %+v\n", roles)
+
+	// REST API returns all project roles, including ones with PREDEFINED type which can't be altered.
+	// We are only interested in the "CUSTOM" types that we can manipulate.
+	customRoles := roles[:0]
+	for _, role := range roles {
+		if role.Type == customRoleType {
```
Comment (chb0github) [inline pkg/projects/resource_project_role.go:160]: You should seperate this into it's own function for completeness sake - it's a filter

File: pkg/projects/resource_project_role.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	return nil
+}
+
+var deleteRole = func(projectKey string, role Role, m interface{}) error {
+	log.Println("[DEBUG] deleteRole")
+	log.Printf("[TRACE] %+v\n", role)
+
+	_, err := m.(*resty.Client).R().Delete(fmt.Sprintf(projectRolesUrl, projectKey) + role.Name)
```
Comment (chb0github) [inline pkg/projects/resource_project_role.go:254]: If you're already using `Sprintf` then also using string concatenation is weird. Keep it consistent: `Sprintf` or `+`

File: pkg/projects/util.go [hunk 6]
```diff
@@ -13,54 +14,27 @@ import (
...
-func (d *ResourceData) getString(key string, onlyIfChanged bool) string {
-	if v, ok := d.GetOk(key); ok && (!onlyIfChanged || d.HasChange(key)) {
+func (d *ResourceData) getString(key string) string {
+	if v, ok := d.GetOk(key); ok {
 		return v.(string)
 	}
 	return ""
 }
 
-func (d *ResourceData) getBoolRef(key string, onlyIfChanged bool) *bool {
```
Comment (chb0github) [inline pkg/projects/util.go:29]: LOL! You got rid of it! Good!

**Review by chb0github — DISMISSED** [review]
Nothing egregious but a few suggestions

File: pkg/projects/resource_project_role.go (hunk 3 above)
Comment (alexhung) [reply pkg/projects/resource_project_role.go:79]: Right?! 😠

File: pkg/projects/resource_project.go (hunk 1 above)
Comment (alexhung) [reply pkg/projects/resource_project.go:201]: Yeah, will update. Interesting you call it gravemark, I always know it as backtick.

File: pkg/projects/resource_project_role.go (hunk 5 above)
Comment (alexhung) [reply pkg/projects/resource_project_role.go:254]: Good idea on using Resty way. Will update.

File: pkg/projects/resource_project_role.go (hunk 2 above)
Comment (alexhung) [reply pkg/projects/resource_project_role.go:65]: I agree. I thought about that this morning and came to similar conclusion

Comment (alexhung) [reply pkg/projects/resource_project_role.go:65]: Having said that, `project` only needs to implement `Identifiable` so I'll keep those 2 interfaces separate for now.

**Review by chb0github — APPROVED** [review]

**Pull request merged**